model/task_queue_service.go

```diff
@@ -0,0 +1,197 @@
+	}
+
+	var unit schedulableUnit
+	var ok bool
+	var next *TaskQueueItem
+
+	if spec.Group != "" {
+		unit, ok = t.units[compositeGroupId(spec.Group, spec.BuildVariant, spec.Version)]
+		if ok {
+			if next = t.nextTaskGroupTask(unit); next != nil {
+				return next
+			}
+		}
+		// If the task group is not present in the task group map, it has been dispatched.
+		// Fall through to getting a task not in that group.
+	}
+	var numHosts int
+	var err error
+	for _, schedulableUnitID := range t.order {
+		unit, ok = t.units[schedulableUnitID]
+		if !ok {
+			continue
+		}
+		// If maxHosts is not set, this is not a task group.
+		if unit.maxHosts == 0 {
+			delete(t.units, schedulableUnitID)
+			return &unit.tasks[0]
+		}
+		if unit.runningHosts < unit.maxHosts {
+			// TODO: For a multi-host task group, it's not possible to correctly
+			// dispatch tasks based on number of hosts running tasks in the task group
+			// without a transaction. When we use a driver that supports transactions,
+			// we likely want to rewrite this code to use transactions. Currently it's
+			// possible to dispatch a multi-host task group to more tasks than max
+			// hosts. Before transactions are available, it might be possible to do this
+			// better by performing this query again, just before returning from this
+			// function.
+			numHosts, err = host.NumHostsByTaskSpec(spec.Group, spec.BuildVariant, spec.ProjectID, spec.Version)
+			if err != nil {
+				grip.Error(message.WrapError(err, message.Fields{
+					"message": "problem running NumHostsByTaskSpec query",
+					"group":   spec.Group,
+					"variant": spec.BuildVariant,
+					"project": spec.ProjectID,
+					"version": spec.Version,
+				}))
+				return nil
+			}
+			unit.runningHosts = numHosts
+			t.units[schedulableUnitID] = unit
+			if unit.runningHosts < unit.maxHosts {
+				if next = t.nextTaskGroupTask(unit); next != nil {
+					return next
+				}
+			}
+		}
+	}
+
+	return nil
+}
+
+func compositeGroupId(group, variant, version string) string {
+	return fmt.Sprintf("%s-%s-%s", group, variant, version)
+}
+
+func (t *TaskDispatchService) nextTaskGroupTask(unit schedulableUnit) *TaskQueueItem {
+	for i, nextTask := range unit.tasks {
+		if nextTask.IsDispatched == true {
+			continue
+		}
+
+		nextTaskFromDB, err := task.FindOneId(nextTask.Id)
+		if err != nil {
+			grip.Error(message.WrapError(err, message.Fields{
+				"message": "problem finding task in db",
+				"task":    nextTask.Id,
+			}))
+			return nil
+		}
+		if nextTaskFromDB == nil {
+			grip.Error(message.Fields{
+				"message": "task from db not found",
+				"task":    nextTask.Id,
+			})
+			return nil
+		}
+
+		// If the task is running in a 1-host task group, has finished, and did not succeed,
+		// remove the units from the local queue. Rely on EndTask to block later tasks.
+		if unit.maxHosts == 1 && nextTaskFromDB.FinishTime != util.ZeroTime && nextTaskFromDB.Status != evergreen.TaskSucceeded {
+			delete(t.units, unit.id)
```

Comment:
Should we also remove it from order?

Reply:
No, see my comment above https://github.com/evergreen-ci/evergreen/pull/2071/files#r262645328.